backend-go/lib.go

```diff
@@ -0,0 +1,229 @@
+	NcJson *nats.EncodedConn
+
+	// DataDir is an absolute path to the data directory of the service.
+	// This is where the service should store any data it needs to persist.
+	// The data is shared between multiple services.
+	// To ensure that the service doesn't overwrite data from other services,
+	// you should create a subdirectory for your service.
+	DataDir string
+
+	// ServiceName is the name of the service.
+	// This is used to create a unique queue group for the service.
+	ServiceName string
+
+	// Config is the assembled configuration from all available configuration sources.
+	// It configures the service.
+	Config *Config
+}
+
+// StartService starts a new Telestion service and returns the available APIs.
+// Additional options passed are applied in the order in they appear and modify the startup procedure.
+//
+// During startup the service parses the service [Config] from different configuration sources in the following order:
+//
+//  1. `overwriteArgs` from [WithOverwriteArgs]
+//  2. command line arguments
+//  3. environment variables
+//  4. default configuration, if `--dev` is passed in the steps from above
+//  5. configuration file, if `CONFIG_FILE` parameter is defined, readable and parsable
+func StartService(opts ...Option) (*Service, error) {
+	options := Options{
+		Nats:          true,
+		OverwriteArgs: map[string]string{},
+		CustomNc:      nil,
+		CustomNcJson:  nil,
+	}
+
+	// run through each option modifier passing the current options
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt(&options); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// load the service configuration
+	config, err := assembleConfig(options.OverwriteArgs)
+	if err != nil {
+		return nil, err
+	}
+
+	// resolve data directory as absolute path
+	dataDir, err := filepath.Abs(config.DataDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// initialize NATS connections
+	// explicitly set to nil if NATS is disabled
+	var nc *nats.Conn = nil
+	var ncJson *nats.EncodedConn = nil
+	if options.Nats {
+		nc, ncJson, err = initializeNats(config, options.CustomNc, options.CustomNcJson)
+		if err != nil {
+			// only close nats connections that aren't externally managed
+			if options.CustomNc == nil && ncJson != nil {
+				nc.Close()
+			}
+			if options.CustomNcJson == nil && nc != nil {
+				ncJson.Close()
+			}
+			return nil, err
+		}
+	}
+
+	return &Service{
+		Nc:          nc,
+		NcJson:      ncJson,
+		DataDir:     dataDir,
+		ServiceName: config.ServiceName,
+		Config:      config,
+	}, nil
+}
+
+// Drain tries to drain both NATS connection of the service.
+func (service *Service) Drain() (errNcJson error, errNc error) {
+	if service.NcJson != nil {
+		errNcJson = service.NcJson.Drain()
+	}
+	if service.Nc != nil {
+		errNc = service.Nc.Drain()
+	}
+
+	return
+}
+
+// Close closes both NATS connections of the service.
+func (service *Service) Close() {
```

Comment:
We don't care about the number 😜
```suggestion
// Close closes any NATS connections of the service.
func (service *Service) Close() {
```